packages/react-native-renderer/src/ReactNativeEventEmitter.js

```diff
@@ -74,123 +77,101 @@ var removeTouchesAtIndices = function(
+ * So two `touchStart`/`touchMove`s that occur simultaneously are received as
+ * two separate touch event dispatches - when they arguably should be one.
+ *
+ * This implementation reuses the `Touch` objects themselves as the `Event`s
+ * since we dispatch an event for each touch (though that might not be spec
+ * compliant). The main purpose of reusing them is to save allocations.
+ *
+ * TODO: Dispatch multiple changed touches in one event. The bubble path
+ * could be the first common ancestor of all the `changedTouches`.
+ *
+ * One difference between this behavior and W3C spec: cancelled touches will
+ * not appear in `.touches`, or in any future `.touches`, though they may
+ * still be "actively touching the surface".
+ *
+ * Web desktop polyfills only need to construct a fake touch event with
+ * identifier 0, also abandoning traditional click handlers.
+ */
+export function receiveTouches(
+  eventTopLevelType: string,
+  touches: Array<Object>,
+  changedIndices: Array<number>,
+) {
+  var changedTouches = eventTopLevelType === 'topTouchEnd' ||
+    eventTopLevelType === 'topTouchCancel'
+    ? removeTouchesAtIndices(touches, changedIndices)
+    : touchSubsequence(touches, changedIndices);
 
-    for (var jj = 0; jj < changedTouches.length; jj++) {
-      var touch = changedTouches[jj];
-      // Touch objects can fulfill the role of `DOM` `Event` objects if we set
-      // the `changedTouches`/`touches`. This saves allocations.
-      touch.changedTouches = changedTouches;
-      touch.touches = touches;
-      var nativeEvent = touch;
-      var rootNodeID = null;
-      var target = nativeEvent.target;
-      if (target !== null && target !== undefined) {
-        if (target < ReactNativeTagHandles.tagsStartAt) {
-          if (__DEV__) {
-            warning(
-              false,
-              'A view is reporting that a touch occurred on tag zero.',
-            );
-          }
-        } else {
-          rootNodeID = target;
+  for (var jj = 0; jj < changedTouches.length; jj++) {
+    var touch = changedTouches[jj];
+    // Touch objects can fulfill the role of `DOM` `Event` objects if we set
+    // the `changedTouches`/`touches`. This saves allocations.
+    touch.changedTouches = changedTouches;
+    touch.touches = touches;
+    var nativeEvent = touch;
+    var rootNodeID = null;
+    var target = nativeEvent.target;
+    if (target !== null && target !== undefined) {
+      if (target < ReactNativeTagHandles.tagsStartAt) {
+        if (__DEV__) {
+          warning(
+            false,
+            'A view is reporting that a touch occurred on tag zero.',
+          );
         }
+      } else {
+        rootNodeID = target;
       }
-      ReactNativeEventEmitter._receiveRootNodeIDEvent(
-        rootNodeID,
-        eventTopLevelType,
-        nativeEvent,
-      );
     }
-  },
-};
-
-export default ReactNativeEventEmitter;
+    // $FlowFixMe Shouldn't we *not* call it if rootNodeID is null?
```

Comment:
I don't think we should call it if `rootNodeID` is `null`, shouldn't we guard against this?

Reply:
Probably should not. I'll file a follow up to ask whoever knows this piece.